link-grammar/parse/prune.c

```diff
@@ -1129,17 +1129,169 @@ static int pp_prune(Sentence sent, Parse_Options opts)
+ * may not have links.
+ * Optional words are totally ignored because they don't have to link.
+ * FIXME: Only the first connector on each side is checked.
+ *
+ * Algo:
+ * 1. For each word (w) find the nearest words (according to the shallow
+ * connectors of its disjuncts) to its left (nl) and right (nr) to * which
+ * a link must exist. I.e., a link from w must exist to nr or a word after
+ * it, or to nl or a word before it.
+ * 2. Using these info, discard disjuncts of words between w and nr, that
+ * must connect to words before w (similarly for nl).
+ */
+static int cross_links_prune(Sentence sent, Parse_Options opts)
+{
+	if (opts->min_null_count > 0) return 0;
+
+	unsigned int *l_nearest_word = alloca(sent->length * sizeof(int));
+	unsigned int *r_nearest_word = alloca(sent->length * sizeof(int));;
+
+	for (unsigned int w = 0; w < sent->length; w++)
+	{
+		l_nearest_word[w] = 0;
+		r_nearest_word[w] = UNLIMITED_LEN;
+
+		for (Disjunct *d = sent->word[w].d; d != NULL; d = d->next)
+		{
+			if (sent->word[w].optional) continue;
+
+			if (NULL == d->left)
+				l_nearest_word[w] = w;
+			else if (d->left->nearest_word > l_nearest_word[w])
+				l_nearest_word[w] = d->left->nearest_word;
+
+			if (NULL == d->right)
+				r_nearest_word[w] = w;
+			else if (d->right->nearest_word < r_nearest_word[w])
+				r_nearest_word[w] = d->right->nearest_word;
+		}
+
+		if (l_nearest_word[w] == 0)
+			l_nearest_word[w] = w;
+		if (r_nearest_word[w] == UNLIMITED_LEN)
+			r_nearest_word[w] = w;
+	}
+
+	Disjunct *free_later = NULL;
+	int total_deleted = 0;
+
+	for (unsigned int w = 0; w < sent->length; w++)
+	{
+		/* Verification-only checks. Maybe they are redundant. */
+		if ((r_nearest_word[w] > sent->length-1) ||
+			 (r_nearest_word[w] < w))
+		{
+			prt_error("Warning: w[%d].nearest_word_right %d\n",
+			          w, r_nearest_word[w]);
+			continue;
+		}
+		if (l_nearest_word[w] > w)
+		{
+			prt_error("Warning: w[%d].nearest_word_left %d\n",
+			          w, l_nearest_word[w]);
+			continue;
+		}
+
+		if (sent->word[w].optional) continue;
+
+		/* Remove disjuncts that are blocked by l->r links. */
+		for (WordIdx rw = w+1; rw < r_nearest_word[w]; rw++)
+		{
+			if (sent->word[rw].optional) continue;
+
+			for (Disjunct **dd = &sent->word[rw].d; *dd != NULL; /* See: NEXT */)
+			{
+				Disjunct *d = *dd; /* just for convenience */
+
+				if ((NULL != d->left) && d->left->nearest_word < w)
+				{
+#if 0
+					printf("w %d nw_right %d rw %d nw_left %d dl %p:",
+					       w, r_nearest_word[w], rw, d->left->nearest_word, d);
+					print_connector_list(d->left);
+					printf("\n");
+#endif
+
+					/* Discard the current disjunct. */
+					*dd = d->next; /* NEXT - set current disjunct to the next one */
+					d->next = free_later;
+					free_later = d;
+					total_deleted++;
+				}
+				else
+				{
+					dd = &d->next; /* NEXT */
+				}
+			}
+		}
+
+		/* Remove disjuncts that are blocked by mandatory r->l links. */
```

Comment:
This code supposes `null_count==0`, so we know that at least one disjunct for each word (which is not optional) must be selected.
Lets look at a word in location `w` > 0, that has disjuncts with left pointing connectors.

Each of all the left pointing shallow connectors has a certain `nearest_word`, and we find the maximum of it. Then we know there must be a link to a word at that position or lower than it (because at least one of the disjuncts must be used).

Now we can apply the constraint that no links may cross it. We examine any word `lw` strictly between the nearest possible left-going link endpoint so computed and `w`. If there is a shallow right-going connector with a `nearest_word > w`, we can discard it because its link must cross the word `w` that must have a left-going link.

This idea can be extended in several ways:
1. Deep connectors constraints can be added. They increase the said linking constraints.
2. A map of ranges that must be "covered" by such minimal links can be created, and any disjunct that has a connector that must connect to a word in such a range can be discarded.
3. For that purpose, the length_limit can also be considered, and also the ability of a word to create a connection to another particular word.
4. A pass can be done over all connectors, extending any `nearest_word` that falls in such a `forbidden` range, and shortening `length_limit`that extends into such a range.
5.`do_count()` can consult a per word map to find out to which other words there is no point to create a link because it will fall into such a "forbidden" range for links from this word.